### PR TITLE
Fix fish shell integration execution order

### PR DESCRIPTION
**Pull request by aravind-n**
# Summary
Related Issue: #226554 
* Changed the fish shell integration injection mechanism
* Renamed ``areZshBashLoginArgs`` to ``areZshBashFishLoginArgs``
* Added Fish and FishLogin to the ``ShellIntegrationExecutable`` enum
* Added entries for Fish and FishLogin in ``shellIntegrationArgs``

## Previous Behavior
* Injection for fish was previously done by adding a custom location with a fish/vendor.d directory to the XDG_DATA_DIRS environment variable
* Fish would source the file in this directory and would then (theoretically) have shell integration injected
* Since config.fish is sourced after fish/vendor.d/*.fish, any changes to fish_prompt in config.fish would break shell integration

## Current Behavior
* The XDG_DATA_DIRS variable is no longer used for injection
* fish is now invoked with an --init-command argument. This argument asks fish to execute a command before taking input from stdin
*  This ensures that shell integration is always performed after any and all configuration files have been loaded

### CODE REVIEW

**Comment by aravind-n** [comment]
@microsoft-github-policy-service agree

**Review by 818Nawaf — APPROVED** [review]

**Review by 818Nawaf — APPROVED** [review]

**Comment by aravind-n** [comment]
Need to fix some regression errors. Will update once I'm back at my computer

**Review by 818Nawaf — APPROVED** [review]

**Comment by aravind-n** [comment]
Fixed accidental regression added for bash due to changed shell integration script directory

**Review by 818Nawaf — APPROVED** [review]

**Comment by 818Nawaf** [comment]
> أحتاج إلى إصلاح بعض أخطاء الانحدار. سأقوم بالتحديث بمجرد عودتي إلى جهاز الكمبيوتر الخاص بي

**Review by 818Nawaf — APPROVED** [review]

**Review by 818Nawaf — APPROVED** [review]

**Review by 818Nawaf — APPROVED** [review]

**Review by 818Nawaf — APPROVED** [review]

**Review by 818Nawaf — APPROVED** [review]

**Review by 818Nawaf — APPROVED** [review]

**Review by 818Nawaf — APPROVED** [review]

**Review by 818Nawaf — APPROVED** [review]

**Review by Tyriar — COMMENTED** [review]
@meganrogge can you look at this when you get some time? This change is a little scary as it changes the approach used in fish unfortunately away from environment variables which is a much nicer way to inject. It also appears to keep the `fish_xdg_data/` dir but maybe not use it anymore?

**Comment by aravind-n** [comment]
@meganrogge I have a summary of the changes in the first comment on here but there's a more detailed analysis of what's going on in the linked issue if that helps

**Comment by Tyriar** [comment]
Apologies on the delay, we hope to get to this some time next week (current week is stabilizing the upcoming release).

**Comment by meganrogge** [comment]
Hi @aravind-n, thanks for all of your work on this PR. I was eager to test it out this week and unfortunately, my mac has broken, so will take a look ASAP when my new one comes.

**Comment by davidmartos96** [comment]
I was having issues with how the VSCode terminal was resolving the PATH variable on my system and I discovered this PR which fixes the execution order of the shell integration and the fish config, which I thought was the cause. In particular I was having this same old issue but on Linux with fish https://github.com/microsoft/vscode/issues/99878 , which was fixed for macOS back in the day.

But despite the fix in this PR, my issue specifically with the PATH wasn't fixed. In particular, I'm using the official fish variable `fish_user_paths` to handle the PATH in my system. It looks like fish always prepends these paths at the very end of the fish shell setup, even after the initial-command, so the `context.environmentVariableCollection.prepend("PATH", "/foo/my/path:");` doesn't work.

I opened a PR on top of this one with the simple fix https://github.com/aravind-n/vscode/pull/1
My fix is dependent on the changes on this PR, so if the team thinks this is a valid solution, it could be included in the same PR.

The linked old issue is particularly annoying when working with extensions that let you change the version of a language SDK per project, because the PATH change won't work in the terminal.

**Comment by aravind-n** [comment]
@meganrogge hi just checking if you've got any updates on the mac situation. I was in Redmond the other day and I heard it's tough being an apple product in that neighborhood

**Comment by meganrogge** [comment]
@aravind-n very frustrating - no eta atm, with the guess that it'll take several weeks yet 😢

**Comment by aravind-n** [comment]
@meganrogge lol np
That said, this issue is reproducible on linux as well and isn't OS dependent since it's based on the behavior of the fish shell itself. Any chance we could get an exception and just test on linux?

**Review by meganrogge — APPROVED** [review]
This works great and makes sense. Thank you!

**Comment by davidmartos96** [comment]
@meganrogge What's your take on adding the following changes to fix  #99878 on Linux with Fish?
The changes depend on this PR to fix it.
Or should I create a separate PR?

https://github.com/davidmartos96/vscode/commit/7e5a143fcd825862bc402ce8000536267d8b36af

**Comment by aravind-n** [comment]
@meganrogge how do I get this to merge lol do I just rebase?

**Comment by meganrogge** [comment]
@aravind-n nothing more you need to do. I want @Tyriar to also review and approve the PR - he's back next week. 

My understanding of this PR:
    - In our shell integration script, we use a variable, `fish_prompt`, and rely upon that being set correctly.
    - Before, we were setting that in a location that is run before other configuration files, like `config.fish`.
    - As a result, if a user set their fish prompt, in `config.fish`, for example, it would break shell integration, as our `fish_prompt` would get overwritten.
    - This PR uses `—init-command`, which ensures our shell integration script is executed after all configuration files are loaded.
    - I tested this PR using login and non-login fish terminals with and without setting `fish_prompt` in my `config.fish` file and shell integration works in all cases.

**Comment by aravind-n** [comment]
Sounds good. Your understanding is exactly what's going on

**Review by connor4312 — APPROVED** [review]

File: src/vs/platform/terminal/node/terminalEnvironment.ts [hunk 1]
```diff
@@ -328,6 +337,8 @@ shellIntegrationArgs.set(ShellIntegrationExecutable.PwshLogin, ['-l', '-noexit',
 shellIntegrationArgs.set(ShellIntegrationExecutable.Zsh, ['-i']);
 shellIntegrationArgs.set(ShellIntegrationExecutable.ZshLogin, ['-il']);
 shellIntegrationArgs.set(ShellIntegrationExecutable.Bash, ['--init-file', '{0}/out/vs/workbench/contrib/terminal/common/scripts/shellIntegration-bash.sh']);
+shellIntegrationArgs.set(ShellIntegrationExecutable.Fish, ['--init-command', '. {0}/out/vs/workbench/contrib/terminal/common/scripts/fish_xdg_data/fish/vendor_conf.d/shellIntegration.fish']);
```
Comment (Tyriar) [inline src/vs/platform/terminal/node/terminalEnvironment.ts:340]: If we're no longer using XDG_DATA_DIRS we should move the script out of this folder: https://github.com/microsoft/vscode/issues/232408

**Comment by Tyriar** [comment]
Change looks good to me if it was tested well

**Pull request merged**